Show changelog on compiler update

nicdard opened the pull request:
Fix: #146

As for now, every time the compiler is updated, a new document (plaintext) is displayed in the editor containing the version of the compiler and the changelog as it is written in the release page. The changelog is displayed as a readonly document thanks to the [virtual document API](https://code.visualstudio.com/api/extension-guides/virtual-documents)

We could discuss whether this behavior is however disturbing for the user because it may interrupt an user reading some code while the extension is loading for example. However it is true that at this point the extension is still loading. To address the potential issue, we could substitute the current behavior with a notification having an open button which when clicked will open the document.

Note: This PR builds on top of https://github.com/flix/vscode-flix/pull/185

magnus-madsen commented:
Question: It only happens when (1) there is a Flix update right and (2) when the extension is loaded? 

If so that would only happen a few times a year-- when we update the latest stable version, so that seems acceptable, don't you think?

nicdard commented:
> Question: It only happens when (1) there is a Flix update right and (2) when the extension is loaded?
> 
> If so that would only happen a few times a year-- when we update the latest stable version, so that seems acceptable, don't you think?

Both yes, so I think it's not an issue, cool!

magnus-madsen commented on client/src/services/releaseVirtualDocument.ts:
```diff
@@ -0,0 +1,84 @@
+        return `Version: ${installedVersion.version.major}.${installedVersion.version.minor}.${installedVersion.version.patch}\n`
+            + `${installedVersion.description}`;
+      } else {
+        throw new Error(`The current installed compiler (${installedVersion.id}) doesn't match the requested one (${id}).`);
```
How can this happen? Do we really want to throw here? 

Does this assume the programmer clicks yes to update the compiler version?

nicdard replied:
No it should never happen, it is only possible if someone later will change the place where the function `openFlixReleaseOverview` is invoked, as for now it should always work as expected

nicdard replied:
I just added it to have better error reporting in case in the future this would happen, but I see your point

magnus-madsen replied:
My fear is that some day this breaks and then the whole extensions breaks. In that case it would better just to not show the changes.

What happens if you run and old version and you decline to update? Do you get the exception then?

nicdard replied:
Nothing happens, since this functionality is only triggers when the installed version is actually set, so just when the download happens successfully

nicdard replied:
Also, I have tried to trigger the exception and the extension code it's not interrupted and just continue the normal execution without showing the CHANGELOG. Indeed the code of the provider is not called by us and is internally triggered by the call to `executeCommand`.

Using a `console.error` would instead let the CHANGELOG be displayed incorrectly with an empty body

magnus-madsen replied:
I am still just concerned about this crashing the extension and making it unusable.

What if we just return "Unable to get latest changelog. Please see XYZ" or something like that?

magnus-madsen commented:
Hmm. Do we have any way of testing this before merge?

magnus-madsen commented:
Oh, and could you show a screenshot?

nicdard commented:
> Hmm. Do we have any way of testing this before merge?

I have been testing it on my own environment, all of the cases:
- download a new version of the compiler because there aren't in the folder: I removed everything in `.config/Code/User/globalStorage/flix.flix`
- download a new release since there is a newer one: I stopped the execution of the code at the point where the check is and changed manually the `downloadAt` field in order to induce this behavior
- no new version is required, do not show: this is the normal behavior

nicdard commented:
> Oh, and could you post a screenshot?

Sure!

nicdard commented:
![Screenshot from 2022-03-06 15-28-28](https://user-images.githubusercontent.com/55001950/156927552-47ad353e-e3f5-4371-b419-4fb0c519070b.png)

magnus-madsen commented:
Would it be possible for us to name the tab? E.g. "Flix Relase Notes" etc?

magnus-madsen commented:
Oh- and what about markdown?

nicdard commented:
I have been trying, but for both of the issues you are mentioning we will need more exploration. By sobstituting `markdown` to `plaintext` the editor recognise the document as markdown, but then you can only see the source (it doesn't open the markdown preview, even when clicking on the button). Also the url is not clickable.

Regarding the name, it seems that it should be possible, but I will need to look better at it since my first tries where failing.

In the meantime I have found those resources:
- https://github.com/microsoft/vscode/issues/102739
- https://code.visualstudio.com/api/references/contribution-points

nicdard commented:
![Screenshot from 2022-03-06 23-11-24](https://user-images.githubusercontent.com/55001950/156944269-b4f99ffe-9793-4455-bbf4-263e54bbf926.png)

nicdard commented:
![Screenshot from 2022-03-07 11-54-34](https://user-images.githubusercontent.com/55001950/157018019-6483889f-0601-490d-92bd-671785f9bab0.png)

I think now it's in the right direction! :muscle:

magnus-madsen commented:
Nice! I like it!

magnus-madsen commented:
I think this is ready. I am just concerned about the exception, but I proposed a solution.

magnus-madsen commented:
Thanks! 

(I merged it, but I cannot deploy it until in about 10 days when I am back... I don't have the deployment token on my laptop AFAIK.)

nicdard commented:
> Thanks!
> 
> (I merged it, but I cannot deploy it until in about 10 days when I am back... I don't have the deployment token on my laptop AFAIK.)

Cool thanks! No rush :)

magnus-madsen commented:
This is now live.

Later this week I will push a new stable version of Flix and we will see if it works :)

nicdard commented:
Cool!

magnus-madsen commented:
Brilliant!:

![image](https://user-images.githubusercontent.com/2287953/159281932-e45caffb-931f-441d-b8a2-543ee5df23e5.png)
